source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1183,6 +1255,25 @@ bool ContextImpl::verifyCertChain(X509& leaf_cert, STACK_OF(X509) & intermediate
   return true;
 }
 
+ValidationResults ContextImpl::customVerifyCertChainForQuic(
+    STACK_OF(X509)& cert_chain, Ssl::ValidateResultCallbackPtr callback, bool is_server,
+    const Network::TransportSocketOptionsConstSharedPtr& transport_socket_options,
+    const CertValidator::ExtraValidationContext& validation_context) {
+  ASSERT(Runtime::runtimeFeatureEnabled("envoy.reloadable_features.tls_async_cert_validation"));
+  ASSERT(!tls_contexts_.empty());
+  // It doesn't matter which SSL context is used, because they share the same cert validation
+  // config.
+  SSL_CTX* ssl_ctx = tls_contexts_[0].ssl_ctx_.get();
+  if (SSL_CTX_get_verify_mode(ssl_ctx) == SSL_VERIFY_NONE) {
+    // Skip validation if the tls is configured SSL_VERIFY_NONE.
```

Comment:
Nit: s/tls/TLS/

Also, does Envoy use SSL_VERIFY_NONE? SSL_VERIFY_NONE for client certs means you don't request a cert anyway, so this would be unreachable. SSL_VERIFY_NONE for server certs is a misfeature and insecure. If Envoy's not using it, it has no reason to emulate it. If Envoy is using it... it should stop.

Reply:
> SSL_VERIFY_NONE can be configured for a Envoy TLS transport socket at either endpoints. The client and the server share the same API.

That sounds like a mistake. Is there a bug filed in Envoy to fix this? It's a false sharing here because `SSL_VERIFY_NONE` doesn't even do the same thing across client and server. If you're a server, `SSL_VERIFY_NONE` means "don't request a client cert". If you're a client, it means "get a server cert (because it's always sent), verify it, make a note of the error, but proceed regardless".

Those aren't the same feature. It's a legacy OpenSSL mistake that they were given the same name. Envoy should not be passing that mistake along to its consumers.

> What we can do is immediate return success upon that callback if the Envoy is configured SSL_VERIFY_NONE.

But that's not what `SSL_VERIFY_NONE` does. `SSL_VERIFY_NONE`, as a client verifying the serve certificate, _still verifies the certificate_. On error, it sticks an error into `SSL_get_verify_result` and proceeds anyway. It's strange for the same mechanism to do different things on the QUIC and TCP side.

Though this mechanism is incoherent so, like I said, the correct behavior is to simply not support it.

Reply:
> > SSL_VERIFY_NONE can be configured for a Envoy TLS transport socket at either endpoints. The client and the server share the same API.
> 
> That sounds like a mistake. Is there a bug filed in Envoy to fix this? It's a false sharing here because `SSL_VERIFY_NONE` doesn't even do the same thing across client and server. If you're a server, `SSL_VERIFY_NONE` means "don't request a client cert". If you're a client, it means "get a server cert (because it's always sent), verify it, make a note of the error, but proceed regardless".
> 
> Those aren't the same feature. It's a legacy OpenSSL mistake that they were given the same name. Envoy should not be passing that mistake along to its consumers.
> 

The shared config API is the [CA path] (https://github.com/envoyproxy/envoy/blob/e19476d62956b60688de2b6109de1cf4dea6ae19/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L332). If the Envoy isn't configured with a CA, it will not call `SSL_CTX_set_verify` and `SSL_CTX_set_cert_verify_callback`, so the SSL_CTX is left with SSL_VERIFY_NONE and no verify callback. My understanding is that neither server nor client will call th Envoy verify callback. How would boring SSL verify the server certs in that case?

Anyway, there is not tracking issue for mixed client and server config API. But I'm not sure what problem this could cause.

> > What we can do is immediate return success upon that callback if the Envoy is configured SSL_VERIFY_NONE.
> 
> But that's not what `SSL_VERIFY_NONE` does. `SSL_VERIFY_NONE`, as a client verifying the serve certificate, _still verifies the certificate_. On error, it sticks an error into `SSL_get_verify_result` and proceeds anyway. It's strange for the same mechanism to do different things on the QUIC and TCP side.
> 

ContextImpl::customVerifyCertChainForQuic() is only used by QUIC client and it is mimicking ContextImpl::verifyCertChain() a few lines above which indeed always verifies and then return. I made the change to early return upon SSL_VERIFY_NONE in customVerifyCertChainForQuic(), thinking it would save some resources. But as you said, the right way to handle SSL_VERIFY_NONE should be actually config the QUIC ssl context to SSL_VERIFY_NONE and verify cert as is, and let the boring stack decide what to do. Is my understanding right?

> Though this mechanism is incoherent so, like I said, the correct behavior is to simply not support it.

Envoy API change probably isn't a good approach.

Reply:
> Instead, it sounds like Envoy's implementation doesn't match the documentation. "If not specified and a peer certificate is presented it will not be verified" does not match what you describe the code as doing. 

The TCP ssl socket does what the documentation says as the SSL custom verify callback is not assigned at all, so the verification is skipped by boringSSL. For QUIC, it is hard-coded to `SSL_VERIFY_PEER` so 
`customVerifyCertChainForQuic()` is always called. But we early return success if the configuration doesn't specify a trusted CA. I was thinking to make QUICHE able to config `SSL_VERIFY_NONE` to align with what TCP does in Envoy, but upon your following comment, it seems that QUICHE behavior is more reasonable. Can you clarify what's the difference you see in the behavior and documentation?

> If you want to skip server certificate verification, you need to configure a callback which accepts every certificate, not `SSL_VERIFY_NONE`. See both the OpenSSL and BoringSSL documentation, which discusses this: https://www.openssl.org/docs/man1.1.1/man3/SSL_set_verify.html https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_VERIFY_NONE

Reply:
I added TODO https://github.com/envoyproxy/envoy/pull/21417/commits/656f9801230cc338bd007ff7dabb53d59faef951 to stop using SSL_VERIFY_NONE in TCP case.